plugins/admin/pkg/command/profiling/downloader.go

```diff
@@ -0,0 +1,179 @@
+	"strconv"
+	"time"
+
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/portforward"
+	"k8s.io/client-go/transport/spdy"
+)
+
+const (
+	pprofPort uint32 = 8008
+)
+
+// ProfileType enums for all supported profiles
+type ProfileType int
+
+const secondsKey = "seconds"
+
+const (
+	ProfileTypeUnknown ProfileType = iota
+	ProfileTypeHeap
+	ProfileTypeProfile
+	ProfileTypeBlock
+	ProfileTypeTrace
+	ProfileTypeAllocs
+	ProfileTypeMutex
+	ProfileTypeGoroutine
+	ProfileTypeThreadCreat
+)
+
+// ProfileEndpoints array maps ProfileType to the string endpoint for pprof
+var ProfileEndpoints = [...]string{
+	ProfileTypeUnknown:     "",
+	ProfileTypeHeap:        "heap",
+	ProfileTypeProfile:     "profile",
+	ProfileTypeBlock:       "block",
+	ProfileTypeTrace:       "trace",
+	ProfileTypeAllocs:      "allocs",
+	ProfileTypeMutex:       "mutex",
+	ProfileTypeGoroutine:   "goroutine",
+	ProfileTypeThreadCreat: "threadcreate",
+}
+
+// DownloadOptions interface to manipulate the http request to pprof server
+type DownloadOptions interface {
+	Apply(*http.Request) error
+}
+
+// Downloader struct holds all private fields
+type Downloader struct {
+	podName   string
+	namespace string
+	// Close this will trigger closing for the endCh create by our own and then cancel all sub goroutines
+	stopCh     <-chan struct{}
+	endCh      chan struct{}
+	readyCh    chan struct{}
+	restConfig *rest.Config
+	localPort  uint32
+	client     *http.Client
+}
+
+// NewDownloader returns the profiling downloader and setup connections asynchronously
+func NewDownloader(cfg *rest.Config, podName, namespace string, endCh <-chan struct{}) (*Downloader, error) {
+	d := &Downloader{
+		podName:    podName,
+		namespace:  namespace,
+		readyCh:    make(chan struct{}),
+		stopCh:     endCh,
+		restConfig: cfg,
+		localPort:  18008,
+		client:     http.DefaultClient,
+	}
+	err := d.connect()
+	if err != nil {
+		return nil, err
+	}
+	return d, nil
+}
+
+// non-blocking call to forward remote port in pod to localhost
+func (d *Downloader) connect() error {
+	path := fmt.Sprintf("/api/v1/namespaces/%s/pods/%s/portforward", d.namespace, d.podName)
+	transport, upgrader, err := spdy.RoundTripperFor(d.restConfig)
+	if err != nil {
+		return err
+	}
+	u, err := url.Parse(d.restConfig.Host)
+	if err != nil {
+		return err
+	}
+	url := &url.URL{
+		Host:   u.Host,
+		Scheme: u.Scheme,
+		Path:   path,
+	}
+	dialer := spdy.NewDialer(upgrader, &http.Client{Transport: transport}, http.MethodPost, url)
+	fw, err := portforward.New(dialer, []string{fmt.Sprintf("%d:%d", d.localPort, pprofPort)}, d.stopCh, d.readyCh, os.Stdout, os.Stderr)
+	if err != nil {
+		return err
+	}
+	go func() {
+		defer close(d.endCh)
+		err := fw.ForwardPorts()
+		// if the func ForwardPorts() returns, the connection should not be available.
+		if err != nil {
+			// TODO: Log for error?
+		}
+	}()
+	return nil
+}
+
+// ProfilingTime is option to add a seconds param in the http request
+type ProfilingTime time.Duration
```

Comment:
Does it make sense to separate this (and methods) into its own file? Might make things easier to navigate and manage in the long run...